File: LeetCode/ReverseInteger/Java/ReverseInteger.java
//
// Created by Shahriar Nasim Nafi on 6/1/22.
//
// Problem: https://leetcode.com/problems/reverse-integer/
//
// A hack with !StringBuilder!
//

class Solution {
    public int reverse(int x) {
        boolean leading = false;
        if (!is32Bit(x)) return 0;
        String X = String.valueOf(x);
        if (X.charAt(0) == '-') {
            leading = true;
           X = X.replace("-","");
        }
        String number = new StringBuilder(String.valueOf(X)).reverse().toString();
        Long result = Long.parseLong(number);
        if (leading) {
            result = result * -1;
        }
        if (!is32Bit(result)) return 0;
        return result.intValue();

    }
    public boolean is32Bit(int num) {
         return (num <= 2147483647) && (num >= -2147483648);
    }
    public boolean is32Bit(Long num) {
         return (num <= 2147483647) && (num >= -2147483648);
    }

}

File: LeetCode/ReverseInteger/CPP/ReverseInteger.cpp
//
// Created by Shahriar Nasim Nafi on 6/1/22.
//
// Problem: https://leetcode.com/problems/reverse-integer/
//
//

class Solution {
public:
    int reverse(int x) {

        if (!(x > INT_MIN && x < INT_MAX)) {
            return 0;
        }


        bool isNegative = false;
        if (x < 0) {
            isNegative = true;
            x = x * -1;
        }

        long long result = 0;
        long digit;
        while (x > 0) {
            digit = (long) x % 10;
            result = result * 10 + digit;
            x /= 10;
        }

        if (isNegative) result = result * -1;
        if (result >= INT_MIN && result <= INT_MAX) {
            return (int) result;
        }
        return 0;
    }
};

